include/mender-update-module.h

```diff
@@ -0,0 +1,132 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef __MENDER_UPDATE_MODULE_H__
+#define __MENDER_UPDATE_MODULE_H__
+
+#include <cJSON.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif /* __cplusplus */
+
+typedef enum mender_update_state_t {
+    MENDER_UPDATE_STATE_DOWNLOAD = 0,
+    MENDER_UPDATE_STATE_INSTALL,
+    MENDER_UPDATE_STATE_REBOOT,
+    MENDER_UPDATE_STATE_VERIFY_REBOOT,
+    MENDER_UPDATE_STATE_COMMIT,
+    MENDER_UPDATE_STATE_CLEANUP,
+    MENDER_UPDATE_STATE_ROLLBACK,
+    MENDER_UPDATE_STATE_ROLLBACK_REBOOT,
+    MENDER_UPDATE_STATE_ROLLBACK_VERIFY_REBOOT,
+    MENDER_UPDATE_STATE_FAILURE,
+    MENDER_UPDATE_STATE_END
+} mender_update_state_t;
+#define N_MENDER_UPDATE_STATES ((size_t)MENDER_UPDATE_STATE_END)
+
+/* The structs below are not supposed to be used directly, only as arguments
+ * passed by in the union below. */
+/**
+ * @param id ID of the deployment
+ * @param artifact name Artifact name
+ * @param type Type from header-info payloads
+ * @param meta_data Meta-data from header tarball
+ * @param filename Artifact filename
+ * @param size Artifact file size
+ * @param data Artifact data
+ * @param index Artifact data index
+ * @param length Artifact data length
+ */
+struct mender_update_download_state_data_s {
+    char    *id;
+    char    *artifact_name;
+    char    *type;
+    cJSON   *meta_data;
+    char    *filename;
+    size_t   size;
+    uint8_t *data;
+    size_t   offset;
+    size_t   length;
+    bool     done;
+};
+
+struct mender_update_install_state_data_s {
+    /* TBD; a NULL pointer in the union below until there is something here */
+};
+
+struct mender_update_reboot_state_data_s {
+    /* TBD; a NULL pointer in the union below until there is something here */
+};
+
+struct mender_update_verify_reboot_state_data_s {
+    /* TBD; a NULL pointer in the union below until there is something here */
+};
+
+struct mender_update_commit_state_data_s {
+    /* TBD; a NULL pointer in the union below until there is something here */
+};
+
+struct mender_update_cleanup_state_data_s {
+    /* TBD; a NULL pointer in the union below until there is something here */
+};
+
+struct mender_update_rollback_state_data_s {
+    /* TBD; a NULL pointer in the union below until there is something here */
+};
+
+struct mender_update_rollback_reboot_state_data_s {
+    /* TBD; a NULL pointer in the union below until there is something here */
+};
+
+struct mender_update_failure_state_data_s {
+    /* TBD; a NULL pointer in the union below until there is something here */
+};
+
+/* The last member allows to type-cast NULL to mender_update_state_data_t,
+ * otherwise we would have to type-cast NULL to one of the other member types
+ * first. */
+/* Should we just use (void *data) as parameter of the callbacks? */
```

Comment:
I don't get the question.

Reply:
Instead of juggling with the union over all the struct pointers when the callback needs to know what to take out from the union anyway. So it might as well just cast a `void *` to whatever it knows is there.